src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/EnumShouldNotHaveDuplicatedValues.cs

```diff
@@ -86,18 +86,18 @@ public sealed override void Initialize(AnalysisContext context)
 
                     var onlyReferencesOneField = GetFilteredDescendants(fieldInitializerOperation, op => op.Kind != OperationKind.Binary)
                         .OfType<IFieldReferenceOperation>()
-                        .Where(fro => fro.Field.ContainingType.Equals(enumSymbol))
-                        .Any();
+                        .Any(fro => fro.Field.ContainingType.Equals(enumSymbol));
                     if (onlyReferencesOneField)
                     {
                         return;
                     }
 
-                    membersByValue.AddOrUpdate(fieldInitializerOperation.Value.ConstantValue.Value,
-                        new ConcurrentBag<(SyntaxNode, string)> { (fieldInitializerOperation.Syntax.Parent, currentField.Name) },
+                    membersByValue.AddOrUpdate(
+                        fieldInitializerOperation.Value.ConstantValue.Value,
+                        new ConcurrentDictionary<(SyntaxNode fieldSyntax, string fieldName), Unit> { [(fieldInitializerOperation.Syntax.Parent, currentField.Name)] = Unit.Default },
                         (key, value) =>
                         {
-                            value.Add((fieldInitializerOperation.Syntax.Parent, currentField.Name));
+                            value.AddOrUpdate((fieldInitializerOperation.Syntax.Parent, currentField.Name), Unit.Default, (key, value) => value);
                             return value;
                         });
```

Comment:
I'm suspicious of this code as the cause of the problem. In general, `AddOrUpdate` allows `updateValueFactory` to get called more than once (in which case the `Add` call would clearly occur too many times). However, it's not clear that this would be possible in the specific case where all calls to `AddOrValue` return `value` unchanged (still trying to verify this).

If I'm correct regarding the underlying problem, perhaps a better fix would be:

* Alter the call to `AddOrUpdate` to `GetOrAdd`
* After obtaining the `ConcurrentBag` from the dictionary, make a single call to `Add` which is guaranteed to only run once

Reply:
🔗 You can see the loop which calls `updateValueFactory` (potentially) multiple times here:
https://github.com/dotnet/runtime/blob/91599fff26619cb8bccb7b39c39590e7cd9a4379/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1276-L1284

The pending verification is whether `TryUpdateInternal` (or `TryUpdate` on .NET Framework) is capable of returning false if values are never modified in the collection after addition.

Reply:
@sharwell How shall we proceed for this PR? Keep `ConcurrentBag` and have a more complex update function or keep going with the current proposed API?

Reply:
I'd prefer the code change to use:

```
ConcurrentDictionary<object, ImmutableHashSet<(SyntaxNode fieldSyntax, string fieldName)>>
```

Reply:
Couple of questions:
- `ImmutableHashSet` is enough because the concurrency is already handled by the `ConcurrentDictionary`?
- Any reason to have immutable data for the value while keeping a mutable data for the outer type?